How to build the project?

This isn't working:

```console
$ git clone https://github.com/httptoolkit/xz-decompress

$ cd xz-decompress

$ npm install

$ docker run -v "$PWD:$PWD" -w "$PWD" -u "$(id -u):$(id -g)" -e wasisdkroot=/opt/wasi-sdk ghcr.io/webassembly/wasi-sdk make
mkdir -p dist/native
"/opt/wasi-sdk/bin/clang" --sysroot="/opt/wasi-sdk/share/wasi-sysroot" \
        --target=wasm32 -DNDEBUG -Os -s -nostdlib -Wl,--no-entry \
        -DXZ_DEC_CONCATENATED -DXZ_USE_CRC64 \
        -Wl,--export=create_context \
        -Wl,--export=destroy_context \
        -Wl,--export=supply_input \
        -Wl,--export=get_next_output \
        -o dist/native/xz-decompress.wasm \
        -Imodule/xz-embedded/userspace/ \
        -Imodule/xz-embedded/linux/include/linux/ \
        module/walloc/walloc.c \
        src/native/*.c \
        module/xz-embedded/linux/lib/xz/xz_crc32.c \
        module/xz-embedded/linux/lib/xz/xz_crc64.c \
        module/xz-embedded/linux/lib/xz/xz_dec_stream.c \
        module/xz-embedded/linux/lib/xz/xz_dec_lzma2.c
src/native/memcmp.c:34:44: error: unknown type name 'size_t'
   34 | void* memmove(void *dest, const void *src, size_t size)
      |                                            ^
src/native/memcmp.c:38:2: error: use of undeclared identifier 'size_t'; did you mean 'sizeof'?
   38 |         size_t i;
      |         ^~~~~~
      |         sizeof
src/native/memcmp.c:38:9: error: use of undeclared identifier 'i'
   38 |         size_t i;
      |                ^
src/native/memcmp.c:41:8: error: use of undeclared identifier 'i'
   41 |                 for (i = 0; i < size; ++i)
      |                      ^
src/native/memcmp.c:41:15: error: use of undeclared identifier 'i'
   41 |                 for (i = 0; i < size; ++i)
      |                             ^
src/native/memcmp.c:41:27: error: use of undeclared identifier 'i'
   41 |                 for (i = 0; i < size; ++i)
      |                                         ^
src/native/memcmp.c:42:6: error: use of undeclared identifier 'i'
   42 |                         d[i] = s[i];
      |                           ^
src/native/memcmp.c:42:13: error: use of undeclared identifier 'i'
   42 |                         d[i] = s[i];
      |                                  ^
src/native/memcmp.c:44:3: error: use of undeclared identifier 'i'
   44 |                 i = size;
      |                 ^
src/native/memcmp.c:45:10: error: use of undeclared identifier 'i'
   45 |                 while (i-- > 0)
      |                        ^
src/native/memcmp.c:46:6: error: use of undeclared identifier 'i'
   46 |                         d[i] = s[i];
      |                           ^
src/native/memcmp.c:46:13: error: use of undeclared identifier 'i'
   46 |                         d[i] = s[i];
      |                                  ^
src/native/memcmp.c:38:2: warning: expression result unused [-Wunused-value]
   38 |         size_t i;
      |         ^~~~~~~~
src/native/memcmp.c:52:30: error: unknown type name 'size_t'
   52 | void* memset(void *s, int c, size_t len) {
      |                              ^
src/native/memcmp.c:65:42: error: unknown type name 'size_t'
   65 | void* memcpy(void *dst, const void *src, size_t len)
      |                                          ^
src/native/memcmp.c:67:5: error: use of undeclared identifier 'size_t'; did you mean 'sizeof'?
   67 |     size_t i;
      |     ^~~~~~
      |     sizeof
src/native/memcmp.c:67:12: error: use of undeclared identifier 'i'
   67 |     size_t i;
      |            ^
src/native/memcmp.c:78:14: error: use of undeclared identifier 'i'
   78 |         for (i = 0; i < len/sizeof(long); i++)
      |              ^
src/native/memcmp.c:78:21: error: use of undeclared identifier 'i'
   78 |         for (i = 0; i < len/sizeof(long); i++)
      |                     ^
src/native/memcmp.c:78:43: error: use of undeclared identifier 'i'
   78 |         for (i = 0; i < len/sizeof(long); i++)
      |                                           ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
1 warning and 20 errors generated.
In file included from src/native/xz-decompress.c:1:
src/native/xz-decompress.h:1:10: fatal error: 'stdlib.h' file not found
    1 | #include <stdlib.h>
      |          ^~~~~~~~~~
1 error generated.
In file included from module/xz-embedded/linux/lib/xz/xz_crc32.c:18:
In file included from module/xz-embedded/linux/lib/xz/xz_private.h:50:
module/xz-embedded/userspace/xz_config.h:40:10: fatal error: 'stdlib.h' file not found
   40 | #include <stdlib.h>
      |          ^~~~~~~~~~
1 error generated.
In file included from module/xz-embedded/linux/lib/xz/xz_crc64.c:13:
In file included from module/xz-embedded/linux/lib/xz/xz_private.h:50:
module/xz-embedded/userspace/xz_config.h:40:10: fatal error: 'stdlib.h' file not found
   40 | #include <stdlib.h>
      |          ^~~~~~~~~~
1 error generated.
In file included from module/xz-embedded/linux/lib/xz/xz_dec_stream.c:10:
In file included from module/xz-embedded/linux/lib/xz/xz_private.h:50:
module/xz-embedded/userspace/xz_config.h:40:10: fatal error: 'stdlib.h' file not found
   40 | #include <stdlib.h>
      |          ^~~~~~~~~~
1 error generated.
In file included from module/xz-embedded/linux/lib/xz/xz_dec_lzma2.c:11:
In file included from module/xz-embedded/linux/lib/xz/xz_private.h:50:
module/xz-embedded/userspace/xz_config.h:40:10: fatal error: 'stdlib.h' file not found
   40 | #include <stdlib.h>
      |          ^~~~~~~~~~
1 error generated.
make: *** [Makefile:23: dist/native/xz-decompress.wasm] Error 1
```

And I tried several other ways too (without docker), and I get the same issue.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -9,10 +9,6 @@ jobs:
       matrix:
         node-version: [16.x, 18.x, 20.x, '*']
 
-    env:
-      WASI_VERSION: 14
-      WASI_SDK_PATH: /tmp/wasi-sdk
-
     steps:
       - uses: actions/checkout@v3
         with:
@@ -25,9 +21,7 @@ jobs:
 
       - name: Set up WASI-SDK
         run: |
-          wget https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-${WASI_VERSION}/wasi-sdk-${WASI_VERSION}.0-linux.tar.gz
-          mkdir -p $WASI_SDK_PATH
-          tar xvf wasi-sdk-*-linux.tar.gz -C $WASI_SDK_PATH --strip-components=1
+          npm run setup-wasi-sdk
 
       - uses: actions/setup-node@v3
         with:
@@ -36,7 +30,7 @@ jobs:
       - run: npm install
 
       - name: Build & test
-        run: wasisdkroot=$WASI_SDK_PATH make && npm run test
+        run: make && npm run test
         env:
           # Legacy provider required for old webpack version in new Node releases:
           NODE_OPTIONS: ${{ matrix.node-version != '16.x' && '--openssl-legacy-provider' || '' }}

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@ sample/lib/
 sample/data/random*
 sample/data/sample.wasm.xz
 sample/data/sample.wasm-brotli.br
+wasi-sdk/

--- a/Makefile
+++ b/Makefile
@@ -2,15 +2,10 @@ xzdir := module/xz-embedded
 xzlibdir := $(xzdir)/linux/lib/xz
 
 ifeq ($(wasisdkroot),)
-  $(error wasisdkroot is not set)
+	wasisdkroot := wasi-sdk
 endif
 
-ifeq ($(shell grep -o WSL2 /proc/version ),WSL2)
-	# Runs a lot faster for me
-	webpackcommand := cmd.exe /c npm run webpack
-else
-	webpackcommand := npm run webpack
-endif
+webpackcommand := node_modules/.bin/webpack
 
 .PHONY: all clean sample run-sample package
 
@@ -64,6 +59,6 @@ run-sample:
 clean:
 	rm -rf dist
 	rm -rf sample/lib
-	rm sample/data/random*
-	rm sample/data/sample.wasm.xz
-	rm sample/data/sample.wasm-brotli.br
+	rm -f sample/data/random*
+	rm -f sample/data/sample.wasm.xz
+	rm -f sample/data/sample.wasm-brotli.br

--- a/README.md
+++ b/README.md
@@ -54,8 +54,8 @@ XZ-Decompress doesn't have built-in support for `.tar`. However, you can use it 
  * Clone/update submodules
     * `git submodule update --init --recursive`
  * Ensure you have a working Clang toolchain that can build wasm
-    * For example, install https://github.com/WebAssembly/wasi-sdk
-    * `export wasisdkroot=/path/to/wask-sdk`
+    * If running on Linux: `npm run setup-wasi-sdk`
+    * Otherwise you can install https://github.com/WebAssembly/wasi-sdk and `export wasisdkroot=/path/to/wasi-sdk`
  * (For testing only) Ensure you have `xz` and `brotli` available as commands on $PATH
  * Run `make`
 

--- a/package.json
+++ b/package.json
@@ -9,6 +9,7 @@
   },
   "scripts": {
     "test": "mocha -r ts-node/register 'test/**/*.spec.ts'",
+    "setup-wasi-sdk": "wasisdkroot=wasi-sdk && wasi_version=21 && rm -rf \"${wasisdkroot}\" && mkdir -p \"${wasisdkroot}\" && curl -fL -# \"https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-${wasi_version}/wasi-sdk-${wasi_version}.0-linux.tar.gz\" | tar -xzf - -C \"${wasisdkroot}\" --strip-components 1",
     "webpack": "webpack"
   },
   "repository": {

--- a/src/xz-decompress.js
+++ b/src/xz-decompress.js
@@ -114,39 +114,57 @@ export class XzReadableStream extends ReadableStream {
             async start(controller) {
                 await XzReadableStream._contextMutex.acquire();
 
-                if (!XzReadableStream._moduleInstance) {
-                    await (XzReadableStream._moduleInstancePromise || (XzReadableStream._moduleInstancePromise = XzReadableStream._getModuleInstance()));
+                try {
+                    if (!XzReadableStream._moduleInstance) {
+                        await (XzReadableStream._moduleInstancePromise || (XzReadableStream._moduleInstancePromise = XzReadableStream._getModuleInstance()));
+                    }
+                    xzContext = new XzContext(XzReadableStream._moduleInstance);
+                } catch (error) {
+                    XzReadableStream._contextMutex.release();
+                    throw error;
                 }
-                xzContext = new XzContext(XzReadableStream._moduleInstance);
             },
 
             async pull(controller) {
-                if (xzContext.needsMoreInput()) {
-                    if (unconsumedInput === null || unconsumedInput.byteLength === 0) {
-                        const { done, value } = await compressedReader.read();
-                        if (!done) {
-                            unconsumedInput = value;
+                try {
+                    if (xzContext.needsMoreInput()) {
+                        if (unconsumedInput === null || unconsumedInput.byteLength === 0) {
+                            const { done, value } = await compressedReader.read();
+                            if (!done) {
+                                unconsumedInput = value;
+                            }
                         }
+                        const nextInputLength = Math.min(xzContext.bufSize, unconsumedInput.byteLength);
+                        xzContext.supplyInput(unconsumedInput.subarray(0, nextInputLength));
+                        unconsumedInput = unconsumedInput.subarray(nextInputLength);
                     }
-                    const nextInputLength = Math.min(xzContext.bufSize, unconsumedInput.byteLength);
-                    xzContext.supplyInput(unconsumedInput.subarray(0, nextInputLength));
-                    unconsumedInput = unconsumedInput.subarray(nextInputLength);
-                }
 
-                const nextOutputResult = xzContext.getNextOutput();
-                controller.enqueue(nextOutputResult.outChunk);
-                xzContext.resetOutputBuffer();
+                    const nextOutputResult = xzContext.getNextOutput();
+                    controller.enqueue(nextOutputResult.outChunk);
+                    xzContext.resetOutputBuffer();
 
-                if (nextOutputResult.finished) {
-                    xzContext.dispose();
+                    if (nextOutputResult.finished) {
+                        xzContext.dispose();
+                        XzReadableStream._contextMutex.release();
+                        controller.close();
+                    }
+                } catch (error) {
+                    if (xzContext) {
+                        xzContext.dispose();
+                    }
                     XzReadableStream._contextMutex.release();
-                    controller.close();
+                    throw error;
                 }
             },
             cancel() {
-                xzContext.dispose();
-                XzReadableStream._contextMutex.release();
-                return compressedReader.cancel();
+                try {
+                    if (xzContext) {
+                        xzContext.dispose();
+                    }
+                    return compressedReader.cancel();
+                } finally {
+                    XzReadableStream._contextMutex.release();
+                }
             }
         });
     }

--- a/src/xz-decompress.js
+++ b/src/xz-decompress.js
@@ -63,9 +63,40 @@ class XzContext {
     }
 }
 
+// Simple mutex to serialize context creation and prevent resource exhaustion
+class ContextMutex {
+    constructor() {
+        this.locked = false;
+        this.waitQueue = [];
+    }
+
+    async acquire() {
+        if (!this.locked) {
+            this.locked = true;
+            return;
+        }
+
+        // Wait in queue
+        return new Promise((resolve) => {
+            this.waitQueue.push(resolve);
+        });
+    }
+
+    release() {
+        if (this.waitQueue.length > 0) {
+            const next = this.waitQueue.shift();
+            next();
+        } else {
+            this.locked = false;
+        }
+    }
+}
+
 export class XzReadableStream extends ReadableStream {
     static _moduleInstancePromise;
     static _moduleInstance;
+    static _contextMutex = new ContextMutex();
+
     static async _getModuleInstance() {
         const base64Wasm = xzwasmBytes.replace('data:application/wasm;base64,', '');
         const wasmBytes = Uint8Array.from(atob(base64Wasm), c => c.charCodeAt(0)).buffer;
@@ -81,6 +112,8 @@ export class XzReadableStream extends ReadableStream {
 
         super({
             async start(controller) {
+                await XzReadableStream._contextMutex.acquire();
+
                 if (!XzReadableStream._moduleInstance) {
                     await (XzReadableStream._moduleInstancePromise || (XzReadableStream._moduleInstancePromise = XzReadableStream._getModuleInstance()));
                 }
@@ -105,12 +138,14 @@ export class XzReadableStream extends ReadableStream {
                 xzContext.resetOutputBuffer();
 
                 if (nextOutputResult.finished) {
-                    xzContext.dispose(); // Not sure if this always happens
+                    xzContext.dispose();
+                    XzReadableStream._contextMutex.release();
                     controller.close();
                 }
             },
             cancel() {
-                xzContext.dispose(); // Not sure if this always happens
+                xzContext.dispose();
+                XzReadableStream._contextMutex.release();
                 return compressedReader.cancel();
             }
         });

--- a/test/test.spec.ts
+++ b/test/test.spec.ts
@@ -47,4 +47,22 @@ describe("Streaming XS decompression", () => {
         expect(result1).to.equal('hello world\n');
         expect(result2).to.equal('hello world\n');
     });
+
+    it("can decompress many demo test streams in parallel", async () => {
+        const streams = [];
+        const promises = [];
+
+        for (let i = 0; i < 10_000; i++) {
+            const dataStream = buildStaticDataStream(Buffer.from(HELLO_WORLD_XZ, 'base64'));
+            const stream = new XzReadableStream(dataStream);
+            streams.push(stream);
+            promises.push(collectOutputString(stream));
+        }
+
+        const results = await Promise.all(promises);
+
+        for (const result of results) {
+            expect(result).to.equal('hello world\n');
+        }
+    });
 });

--- a/test/test.spec.ts
+++ b/test/test.spec.ts
@@ -65,4 +65,81 @@ describe("Streaming XS decompression", () => {
             expect(result).to.equal('hello world\n');
         }
     });
+
+    it("handles errors without causing deadlocks", async () => {
+        // Create a stream that will fail during decompression
+        const badDataStream = buildStaticDataStream(Buffer.from('invalid-xz-data', 'utf8'));
+        const stream = new XzReadableStream(badDataStream);
+
+        // The stream should fail, but not hang
+        try {
+            await collectOutputString(stream);
+            expect.fail('Expected stream to throw an error');
+        } catch (error) {
+            // This is expected - the stream should fail with an error
+            expect(error).to.be.an('error');
+        }
+
+        // After the error, we should be able to create new streams successfully
+        // This verifies that the mutex was properly released
+        const validDataStream = buildStaticDataStream(Buffer.from(HELLO_WORLD_XZ, 'base64'));
+        const validStream = new XzReadableStream(validDataStream);
+
+        const result = await collectOutputString(validStream);
+        expect(result).to.equal('hello world\n');
+    });
+
+    it("handles stream cancellation without deadlocks", async () => {
+        const dataStream = buildStaticDataStream(Buffer.from(HELLO_WORLD_XZ, 'base64'));
+        const stream = new XzReadableStream(dataStream);
+
+        // Start reading the stream
+        const reader = stream.getReader();
+        const { value } = await reader.read();
+        expect(value).to.be.instanceOf(Uint8Array);
+
+        // Cancel the stream
+        await reader.cancel();
+
+        // After cancellation, we should be able to create new streams successfully
+        // This verifies that the mutex was properly released during cancellation
+        const validDataStream = buildStaticDataStream(Buffer.from(HELLO_WORLD_XZ, 'base64'));
+        const validStream = new XzReadableStream(validDataStream);
+
+        const result = await collectOutputString(validStream);
+        expect(result).to.equal('hello world\n');
+    });
+
+    it("handles multiple errors without permanent deadlocks", async () => {
+        const promises = [];
+
+        // Create multiple streams that will fail
+        for (let i = 0; i < 5; i++) {
+            const badDataStream = buildStaticDataStream(Buffer.from(`invalid-xz-data-${i}`, 'utf8'));
+            const stream = new XzReadableStream(badDataStream);
+            promises.push(
+                collectOutputString(stream).catch(error => {
+                    // Expected to fail
+                    expect(error).to.be.an('error');
+                    return null; // Mark as handled
+                })
+            );
+        }
+
+        // Wait for all streams to fail
+        const results = await Promise.all(promises);
+
+        // All should have failed (returned null from catch)
+        for (const result of results) {
+            expect(result).to.be.null;
+        }
+
+        // After all errors, we should still be able to create valid streams
+        // This verifies that the mutex isn't permanently locked
+        const validDataStream = buildStaticDataStream(Buffer.from(HELLO_WORLD_XZ, 'base64'));
+        const validStream = new XzReadableStream(validDataStream);
+
+        const result = await collectOutputString(validStream);
+        expect(result).to.equal('hello world\n');
+    });
 });

